DOCSP-41769: Improved bulk write API (#590)
mongodb/docs-java: source/fundamentals/crud/write-operations/bulk.txt

   :depth: 2
   :class: singlecol

.. meta::
   :keywords: insert, update, replace, code example, efficiency

Overview
--------

In this guide, you can learn how to use bulk operations in the
MongoDB Java Driver. 

To perform a single create, replace, update, or delete operation, you can use
the corresponding method. For example, to insert one document and replace one 
document, you can use the ``insertOne()`` and ``replaceOne()`` methods. In this
case, the ``MongoClient`` makes a call to the database for each operation. 

Generally, you can reduce the number of calls to the database by using bulk write 
operations. You can perform bulk write operations at the following levels:

- :ref:`Collection Level <java-sync-coll-bulk-write>`: You can use the 
  ``MongoCollection.bulkWrite()`` method to perform bulk write operations on a 
  single collection. This method groups write operations together by the kind
  of operation. For example, ``MongoCollection.bulkWrite()`` puts multiple update
  operations in the same call, but makes two calls to the database for an insert

>   operation and a replace operation.

Author: I left mostly as is because I wasn't sure if this made it clearer - if you feel strongly about this I can change, also happy to discuss further

>   of operation. For example, ``MongoCollection.bulkWrite()`` puts multiple insert
>   operations, including ``insertOne()`` and ``insertMany()``, in the same call, 
>   but will make two calls to the database for an insert operation and a replace 
>   operation.

Reviewer: The `insertOne()` and `insertMany()` methods here may refer either to the `MongoCollection` methods, or to the `ClientNamespacedWriteModel` methods, but there are no such methods in the collection bulk write API. We should refer to the proper program elements (note that there is no `InsertManyModel`):

```suggestion
  of operation. For example, ``MongoCollection.bulkWrite()`` puts multiple insert
  operations, including ``UpdateOneModel`` and ``UpdateManyModel``, in the same call, 
  but will make two calls to the database for an `UpdateOneModel` a `ReplaceOneModel`.
```

> - :ref:`Client Level <java-sync-client-bulk-write>`: When running {+mdb-server+} 
>   version 8.0 or later, you can use the ``MongoClient.bulkWrite()`` method to perform 

Reviewer 2: ```suggestion
- :ref:`Client Level <java-sync-client-bulk-write>`: If your application connects to 
   {+mdb-server+} version 8.0 or later, you can use the ``MongoClient.bulkWrite()``
   method to perform 
```

  bulk write operations on multiple collections and databases in the same cluster.

>   This method groups multiple kinds of write operations into one call.

Author: Re-review here

>   This method groups multiple kinds of write operations into one call.

Reviewer 2: ```suggestion
  This method performs all write operations in one database call.
```

.. _java-sync-coll-bulk-write:

Collection Bulk Write
---------------------

> Bulk operations consist of many write operations. To perform a bulk operation 

Reviewer 2: S: specify bulk write

```suggestion
Bulk write operations contain one or more write operations. To perform a bulk write operation 
```

at the collection level, pass a ``List`` of ``WriteModel`` documents to the 
``MongoCollection.bulkWrite()`` method. A ``WriteModel`` is a model that 

> represents any of the write operations. 

Reviewer 2: ```suggestion
represents a write operation. 
```

> different batches. For example, when the method is passed an ``insertOne()``, 
> ``insertMany()``, and ``replaceOne()`` operation, it will split the insert operations

Reviewer: Similarly to https://github.com/mongodb/docs-java/pull/590#discussion_r1880825918:

```suggestion
different batches. For example, when the method is passed an ``DeleteOneModel``, 
``DeleteManyModel``, and ``ReplaceOneModel`` operation, it will split the delete operations
```

> The ``MongoCollection.bulkWrite()`` method splits operations of different kinds into
> different batches. For example, when you pass ``DeleteOneModel``, 

Reviewer 2: S:

```suggestion
The ``MongoCollection.bulkWrite()`` method performs each kind of write operation
in a separate batch. For example, when you pass ``DeleteOneModel``, 
```

> the client may reorder operations for efficiency.

Reviewer: Let's add something like "if the bulk operation is not ordered", with "not ordered" being a link to the "Order of Execution" sub-section of the "Collection Bulk Write" section.

Author: I left out the link because of docs team link practices.

> ``DeleteManyModel``, and ``ReplaceOneModel`` operations to the method, the method

Reviewer 2: ```suggestion
``DeleteManyModel``, and ``ReplaceOneModel`` objects to the method, the method
```

> ``DeleteManyModel``, and ``ReplaceOneModel`` operations to the method, the method

Reviewer 2: ```suggestion
``DeleteManyModel``, and ``ReplaceOneModel`` operations to the method, it
```

> splits the delete operations into one batch and the replace operation 
> into another. During this process, the client might reorder operations for 

Reviewer 2: S:

```suggestion
creates two batches: one for the delete operations and one for the replace operation.
During this process, the client might reorder operations for 
```

between these two sentences, maybe mention that a batch = one database call (unless that's somewhere else on the page)

> efficiency if the bulk operation is not ordered.

Author: Re-review here

> efficiency if the bulk operation is not ordered.

Reviewer 2: S: put this last line in a note admonition and link to the section on ordering

In this guide, you can learn how to use bulk operations in the
MongoDB Java Driver. 

To perform a single create, replace, update, or delete operation, you can use
the corresponding method. For example, to insert one document and replace one 

> document, you can use the ``insertOne()`` and ``replaceOne()`` methods. The 

Reviewer 3: ```suggestion
document, you can use the ``insertOne()`` and ``replaceOne()`` methods. In this case, your 
```

``MongoClient`` makes a call to the database for each operation. You can reduce
the number of calls to one by using bulk write operations.

You can perform bulk write operations by using the bulk write API in the

> {+driver-short+} driver to perform multiple data changes in one call. You can 

Reviewer 3: ```suggestion
{+driver-short+} to perform multiple data changes in one call. You can 
```

> {+driver-short+} to perform multiple data changes in one call. You can 

Reviewer: The new wording

> However, by using bulk write operations, you can reduce the number of calls.

Is still written such that it applies to the specific situation described before it, and still gives the reader a false impression that a bulk write reduces the number of calls in that particular situation.

I think, we need to be honest and not be intentionally obscure. Let's say something like (I am only trying to give an idea of what we should say, not how):

> Generally, you can reduce the number of calls by by using bulk write operations. In the aforementioned scenario, a collection bulk write would not reduce the number of calls, while a client bulk write would. See <an anchor link to Collection Bulk Write> and <an anchor link to Client Bulk Write> for more details.

And then in each of the aforementioned sections we describe how they behave when facing different kinds of write models (insert, delete, update, replace): client bulk writes don't care about different kinds, while collection bulk writes can only pack write models of the same kind in a batch, so a bulk write is split into batches when different kinds of write models are used. Such splitting is more efficient for unordered collection bulk writes, as the client may reorder write models when grouping them in batches.

> {+driver-short+} to perform multiple data changes in one call. You can 

Author: Changed. I used the word "call" instead of "batch" for readability, but I suspect this may be technically incorrect - please let me know

perform bulk operations at the following levels:

- :ref:`Collection Level <java-sync-coll-bulk-write>`: You can use the 

>   ``MongoCollection.bulkWrite()`` API to perform bulk write operations on a 

Reviewer 3: ```suggestion
  ``MongoCollection.bulkWrite()`` method to perform bulk write operations on documents in a single
```

  collection.

> - :ref:`Client Level <java-sync-client-bulk-write>`: In {+mdb-server+} version 
>   8.0 or later, you can use the ``MongoClient.bulkWrite()`` API to perform bulk write 
>   operations on multiple collections and databases in the same cluster.

Reviewer 3: ```suggestion
- :ref:`Client Level <java-sync-client-bulk-write>`: When running {+mdb-server+} version 
  8.0 or later, you can use the ``MongoClient.bulkWrite()`` method to perform bulk write 
  operations on multiple collections and databases in the same cluster.
```

In this guide, you can learn how to use bulk operations in the
MongoDB Java Driver. 

.. note:: Client Bulk Write Method

   This guide primarily focuses on the ``MongoCollection.bulkWrite()``
   method, which allows you to perform bulk operations on a collection. 
   When connecting to a deployment running {+mdb-server+} 8.0 or later,
   you can use the ``MongoClient.bulkWrite()`` method to write
   to multiple databases and collections in the same call. See the 
   :ref:`java-sync-client-bulk-write` section below to learn how to use the

>    ``MongoClient.bulkWrite()`` method.

Reviewer 3: S: on second thought, since this guide covers both types of bulk writes, maybe you can just remove this note and change the content of the overview instead. For example, you can revise the overview section to say something like 
"You can perform bulk write operations by using the Java driver to perform multiple data changes in one call. You can perform bulk operations at the following levels:

- collection level: xxxx
- client level: xxxx

And then add some information about the apis and link to the appropriate sections

- `ordered() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/BulkWriteOptions.html#ordered(boolean)>`__

.. _java-sync-client-bulk-write:

Client Bulk Write Method
------------------------

When connecting to a deployment running {+mdb-server+} 8.0 or later,
you can use the ``MongoClient.bulkWrite()`` method to write
to multiple databases and collections in the same cluster.

In the examples in preceding sections of this page, the ``MongoCollection.bulkWrite()`` method 
takes a list of ``WriteModel`` instances in which the specified subclass of 
``WriteModel`` represents the corresponding write operation. For example, an
instance of ``InsertOneModel`` represents an operation to insert one document. 

Similarly, the ``MongoClient.bulkWrite()`` method takes a 
list of ``ClientNamespacedWriteModel`` instances to represent different write operations.
However, you do not need to specify the subclass that represents the corresponding
write operation. Instead, the ``ClientNamespacedWriteModel`` object contains different 
methods to represent different write operations, such as ``ClientNamespacedWriteModel.insertOne()`` or 

> ``ClientNamespacedWriteModel.updateOne()``.

Reviewer 3: Great job editing this section!

These methods take a ``MongoNamespace`` object that defines which
database and collection to write to, and a ``Document`` object that defines
information about the write operation. Some methods, such as ``updateOne()`` and 

> ``replaceOne()``, also take a ``Filters`` object that defines the filter for the operation.

Reviewer 3: ```suggestion
``replaceOne()``, also take a ``Filters`` object to make changes to only a subset of documents.
```

The following sections describe how to use the client ``bulkWrite()`` method.

Insert Example
~~~~~~~~~~~~~~

> This following example shows how to use the ``bulkWrite()`` method to insert 

Reviewer 3: ```suggestion
This example shows how to use the ``bulkWrite()`` method to insert 
```

> two documents. One document is inserted into the ``people`` collection in a
> database named ``db``. The other document is inserted into a collection called ``things``
> in the ``db`` database.

Reviewer 3: ```suggestion
two documents. One document is inserted into the ``db.people`` collection, while the other document is inserted into the ``db.things`` collection.
```

We use the ``MongoNamespace`` object to define the database and collection we

> want each write operation to be applied to.

Reviewer 3: Avoid we - https://www.mongodb.com/docs/meta/style-guide/writing/write-to-the-user/

.. code-block:: java

   MongoNamespace peopleNamespace = new MongoNamespace("db", "people");
   MongoNamespace thingsNamespace = new MongoNamespace("db", "things");

   List<ClientNamespacedWriteModel> bulkOperations = new ArrayList<>();

>    bulkOperations.add(ClientNamespacedWriteModel.insertOne(peopleNamespace, new Document("name", "Julia Smith")));
>    bulkOperations.add(ClientNamespacedWriteModel.insertOne(thingsNamespace, new Document("object", "washing machine")));

Reviewer 3: S: newline where possible to prevent horizontal scroll

```suggestion
   bulkOperations.add(ClientNamespacedWriteModel.insertOne(
       peopleNamespace,
       new Document("name", "Julia Smith")
   ));
   bulkOperations.add(ClientNamespacedWriteModel.insertOne(
       thingsNamespace, 
       new Document("object", "washing machine")
   ));
```

   ClientBulkWriteResult result = client.bulkWrite(bulkOperations);

.. TODO: link documentation

Replace Example
~~~~~~~~~~~~~~~

The following example shows how to use the ``bulkWrite()`` method to replace

> existing documents in the ``people`` and ``things`` collections.

Reviewer 3: ```suggestion
existing documents in the ``people`` and ``things`` collections of the ``db`` database.
```

Author: changed to "db.people" and "db.things" for consistency with insert section

.. code-block:: java

   MongoNamespace peopleNamespace = new MongoNamespace("db", "people");
   MongoNamespace thingsNamespace = new MongoNamespace("db", "things");

   List<ClientNamespacedWriteModel> bulkOperations = new ArrayList<>();

   bulkOperations.add(ClientNamespacedWriteModel.replaceOne(
         peopleNamespace, 
         Filters.eq("_id", 1),
         new Document("name", "Maximus Farquaad")
      )
   );

   bulkOperations.add(ClientNamespacedWriteModel.replaceOne(
         thingsNamespace, 
         Filters.eq("_id", 1),
         new Document("object", "potato")
      )
   );

   ClientBulkWriteResult result = client.bulkWrite(bulkOperations);

After this example runs successfully, the document in the ``people`` collection in 
which the ``_id`` field has the value of ``1`` is replaced with a new
document. The document in the ``things`` collection in which the 
``_id`` field has the value of ``1`` is replaced with a new document.

.. _java-sync-client-bulk-write-options:

Bulk Write Options
~~~~~~~~~~~~~~~~~~

You can use the ``ClientBulkWriteOptions`` interface to specify options when
running the ``bulkWrite()`` method.

Order of Execution Example

> ``````````````````````````

Author: leaving for consistency with other sections

> By default, the write operations execute in the order that you specify them. However,

Reviewer 3: ```suggestion
By default, the driver performs write operations in the order that you specify them. However,
```

you can pass ``false`` to the ``ordered()`` method on the ``ClientBulkWriteOptions``
interface to perform write operations in an unordered way. When using the unordered
option, an error-producing operation will not block execution of other bulk 
write operations.

The following code shows how to set the ``ordered()`` method when creating an

> instance of ``ClientBulkWriteOptions``.

Reviewer 3: ```suggestion
instance of ``ClientBulkWriteOptions`` and then performs a bulk operation to insert multiple documents.
```

.. code-block:: java

   MongoNamespace namespace = new MongoNamespace("db", "people");

   ClientBulkWriteOptions options = new ClientBulkWriteOptions().ordered(false);

   List<ClientNamespacedWriteModel> bulkOperations = new ArrayList<>();

>    bulkOperations.add(ClientNamespacedWriteModel.insertOne(namespace, new Document("_id", 1).append("name", "Donkey Kong)));

Reviewer 3: S: same comment as above - here are some name guidelines https://www.mongodb.com/docs/meta/style-guide/style/names/

>    // duplicate key

Reviewer 3: ```suggestion
   // Causes a duplicate key error 
```

   bulkOperations.add(ClientNamespacedWriteModel.insertOne(namespace, new Document("_id", 1).append("name", "Mario")));

   bulkOperations.add(ClientNamespacedWriteModel.insertOne(namespace, new Document("name", "Princess Peach")));

   ClientBulkWriteResult result = client.bulkWrite(bulkOperations, options);

> Even though the write operation inserting a document with a duplicate key will
> result in an error, the other operations will be executed because we have specified
> the ``ordered()`` option to be ``false``.

Reviewer 3: ```suggestion
Even though the write operation to insert a document that a duplicate key
results in an error, the other operations are committed because the operation is unordered.
```

For more information about the methods and classes mentioned in this section,
see the following API documentation:

- `BulkWriteOptions <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/BulkWriteOptions.html>`__
- `ordered() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/BulkWriteOptions.html#ordered(boolean)>`__

.. _java-sync-client-bulk-write:

Client Bulk Write
-----------------

When connecting to a deployment running {+mdb-server+} 8.0 or later,
you can use the ``MongoClient.bulkWrite()`` method to write
to multiple databases and collections in the same cluster.

In the examples in preceding sections of this page, the ``MongoCollection.bulkWrite()`` method 
takes a list of ``WriteModel`` instances in which the specified subclass of 
``WriteModel`` represents the corresponding write operation. For example, an
instance of ``InsertOneModel`` represents an operation to insert one document. 

Similarly, the ``MongoClient.bulkWrite()`` method takes a 
list of ``ClientNamespacedWriteModel`` instances to represent different write operations.
However, you do not have to specify the subclass that represents the corresponding

> write operation. Instead, the ``ClientNamespacedWriteModel`` object contains different 

Reviewer 3: ```suggestion
Instead of selecting a specific ``WriteModel`` subclass for each type of write operation,
you can select from the ``ClientNamespacedWriteModel``
```

> methods to represent different write operations, such as ``ClientNamespacedWriteModel.insertOne()`` or 

Reviewer 3: ```suggestion
methods to perform different write operations, such as ``ClientNamespacedWriteModel.insertOne()`` or 
```

Author: will keep as represent since writemodel doesn't necessarily perform

         new Document("name", "Frederic Hilbert")
      )
   );

   bulkOperations.add(ClientNamespacedWriteModel.replaceOne(
         thingsNamespace, 
         Filters.eq("_id", 1),
         new Document("object", "potato")
      )
   );

   ClientBulkWriteResult result = mongoClient.bulkWrite(bulkOperations);

After this example runs successfully, the document that has an ``_id`` value of ``1`` 
in the ``people`` collection is replaced with a new document. The document in 
the ``things`` collection that has an ``_id`` value of ``1`` 
is replaced with a new document.

.. _java-sync-client-bulk-write-options:

Bulk Write Options
~~~~~~~~~~~~~~~~~~

You can pass an instance of ``ClientBulkWriteOptions`` to the ``bulkWrite()``
method to specify options when running the bulk write operations.

Order of Execution Example
``````````````````````````

By default, the driver performs write operations in the order that you specify them until
an error occurs, or until they execute successfully. However, you can pass 
``false`` to the ``ordered()`` method on the ``ClientBulkWriteOptions``
interface to perform write operations in an unordered way. When using the unordered

> option, an error-producing operation does not block execution of other bulk 
> write operations.

Reviewer 3: ```suggestion
option, an error-producing operation does not block execution of other 
write operations in the bulk operation.
```

The following code sets the ``ordered()`` method on an
instance of ``ClientBulkWriteOptions`` and performs a bulk write operation to 
insert multiple documents.

.. code-block:: java

   MongoNamespace namespace = new MongoNamespace("db", "people");

   ClientBulkWriteOptions options = new ClientBulkWriteOptions().ordered(false);

   List<ClientNamespacedWriteModel> bulkOperations = new ArrayList<>();

>    bulkOperations.add(ClientNamespacedWriteModel.insertOne(namespace, new Document("_id", 1).append("name", "Rudra Suraj)));

Reviewer 3: S: missing quote is causing rendering error
```suggestion
   bulkOperations.add(ClientNamespacedWriteModel.insertOne(namespace, new Document("_id", 1).append("name", "Rudra Suraj")));
```

   
   // Causes a duplicate key error
   bulkOperations.add(ClientNamespacedWriteModel.insertOne(namespace, new Document("_id", 1).append("name", "Mario Bianchi")));

   bulkOperations.add(ClientNamespacedWriteModel.insertOne(namespace, new Document("name", "Wendy Zhang")));

>    ClientBulkWriteResult result = mongoClient.bulkWrite(bulkOperations, options);

Reviewer 3: I: something is causing the rendering to appear incorrectly here.

- `ordered() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/BulkWriteOptions.html#ordered(boolean)>`__

.. _java-sync-client-bulk-write:

Client Bulk Write
-----------------

When connecting to a deployment running {+mdb-server+} 8.0 or later,
you can use the ``MongoClient.bulkWrite()`` method to write
to multiple databases and collections in the same cluster.

In the examples in preceding sections of this page, the ``MongoCollection.bulkWrite()`` method 
takes a list of ``WriteModel`` instances in which the specified subclass of 
``WriteModel`` represents the corresponding write operation. For example, an
instance of ``InsertOneModel`` represents an operation to insert one document. 

Similarly, the ``MongoClient.bulkWrite()`` method takes a 
list of ``ClientNamespacedWriteModel`` instances to represent different write operations.
Instead of selecting a specific subclass for each write operation, the 
``ClientNamespacedWriteModel`` object contains different methods to represent 
write operations. ``ClientNamespacedWriteModel`` contains the methods 
``insertOne()``, ``updateOne()``, ``updateMany()``, ``replaceOne()``, 

> ``deleteOne()``, and ``deleteMany()``.

Author: do you mean calling a constructor for the collection-level API?

> ``deleteOne()``, and ``deleteMany()``.

Reviewer: Yes, sorry. I updated the first comment in this conversation: "by calling a constructor for the _client_-level API" -> "by calling a constructor for the _collection_-level API".

I am OK with the new wording in `bulk.txt`.

These methods take a ``MongoNamespace`` object that defines which

> database and collection to write to, and a ``Document`` object that defines

Author: While they do not specifically accept the `org.bson.Document` type, I tried to highlight that users *can* use this type to define a document to be inserted, for the sake of common examples and for consistency with the existing documentation on `MongoCollection.bulkWrite()` on this page. When the API docs are available to be linked, I will link them so users can see all types that are accepted by these methods.

I tried to reword this section to reflect that the provided types can be, but aren't exclusively, used by these methods. Please let me know if you feel strongly about this or any information is incorrect.

Starting with {+mdb-server+} version 8.0 and {+driver-short+} version 5.3, there
is an improved bulk write method, ``MongoClient.bulkWrite()``, that writes to 
different databases and collections in the same cluster.

In the previous examples on this page, the ``bulkWrite()`` method takes a list
of ``WriteModel`` documents in which the specified subclass of ``WriteModel`` 
represents the corresponding write operation. For example, ``InsertOneModel`` 
represents inserting one document. 

Similarly, the new ``bulkWrite()`` method takes a 
list of ``ClientNamespacedWriteModel`` objects to represent the write operation. 
However, you do not need to specify the subclass that represents the corresponding
write operation. Instead, the ``ClientNamespacedWriteModel`` object contains different 
methods to represent different write operations, such as ``insertOne()`` or 
``updateOne()``. 

These methods take a ``MongoNamespace`` object that defines which
database and collection to write to, and a ``Document`` object that defines the
document information. Some methods, such as ``updateOne()`` and ``replaceOne()``,
also take a ``Filters`` object that defines the filter for the operation.

> The following sections describe how to use the new ``bulkWrite()`` method.

Reviewer 3: S: comb the PR for "new" and remove

Insert Example
~~~~~~~~~~~~~~

> The following example shows how to use the new ``bulkWrite()`` method to insert 

Reviewer 3: ```suggestion
This following example shows how to use the new ``bulkWrite()`` method to insert 
```

two documents. One document is inserted into the ``people`` collection in our
database. The other document is inserted into a collection called ``things``.
We use the ``MongoNamespace`` object to define the database and collection we

> want each write operation to be applied to.

Reviewer 3: S: include the database name - saying "our" is ambiguous

.. code-block:: java

   MongoNamespace peopleNamespace = new MongoNamespace("db", "people");
   MongoNamespace thingsNamespace = new MongoNamespace("db", "things");

   ClientBulkWriteResult result = client.bulkWrite(List<>(
      ClientNamespacedWriteModel.insertOne(peopleNamespace, new Document("name", "Julia Smith")),
      ClientNamespacedWriteModel.insertOne(thingsNamespace, new Document("object", "washing machine"))
   ));

After this example is executed, the ``people`` collection holds a 
``{ "name": "Julia Smith" }`` document and the ``things`` collection holds

> a ``{ "object": "washing machine" }`` document.

Reviewer 3: S: i dont think that this sentence is necessary

> .. TODO: link documentation

Reviewer 3: S: what documentation are you linking here?

Replace Example
~~~~~~~~~~~~~~~

> The following example shows how to use the ``bulkWrite()`` method to replace an
> existing document in the ``people`` collection and an existing document in the 
> ``things`` collection.

Reviewer 3: ```suggestion
The following example shows how to use the ``bulkWrite()`` method to replace
existing documents in the ``people`` and ``things`` collections.
```

.. code-block:: java

   MongoNamespace peopleNamespace = new MongoNamespace("db", "people");
   MongoNamespace thingsNamespace = new MongoNamespace("db", "things");

   ClientBulkWriteResult result = client.bulkWrite(List<>(
      ClientNamespacedWriteModel.replaceOne(
         peopleNamespace, 
         Filters.eq("_id", 1),
         new Document("name", "Maximus Farquaad")
      ),
      ClientNamespacedWriteModel.replaceOne(
         thingsNamespace, 
         Filters.eq("_id", 1),
         new Document("object", "potato")
      )
   ));

> After this example is executed, the document in the ``people`` collection in 
> which the ``_id`` field has the value of ``1`` has been replaced with a new
> ``{ "name": "Maximus Farquaad" }`` document. Also, the document
> in the ``things`` collection in which the ``_id`` field has the value of ``1``
> has been replaced with a ``{ "object": "potato" }`` document.

Reviewer 3: ```suggestion
After this example runs successfully, the document in the ``people`` collection in 
which the ``_id`` field has the value of ``1`` is replaced with a new document. The document
in the ``things`` collection in which the ``_id`` field has the value of ``1``
is replaced with a new document.
```

Bulk Write Options
~~~~~~~~~~~~~~~~~~

> Similarly to the :ref:`orderOfExecution` section above, you can use the 
> ``ClientBulkWriteOptions`` object to execute the new ``bulkWrite()`` method
> with options.

Reviewer 3: ```suggestion
You can create a ``ClientBulkWriteOptions`` instance to specify options when running the ``bulkWrite()`` method.
```

Order of Execution Example
``````````````````````````

> As described in the previous Order of Execution section, by default the write 
> operations execute in the order they're specified. However, we can pass 
> ``false`` to the ``ordered()`` method on the ``ClientBulkWriteOptions`` object 
> to execute write operations in any order.

Reviewer 3: S: remove the cross link to the other section. 
```suggestion
By default, the write operations execute in the order that you specify them. However, you can pass 
``false`` to the ``ordered()`` method on the ``ClientBulkWriteOptions`` object 
to perform write operations in an unordered way, so that an error-producing operation does not prevent
other operations from being attempted.
```

> The following code shows how to use the ``ordered()`` method on the 
> ``ClientBulkWriteOptions`` object.

Reviewer 3: ```suggestion
The following code shows how to set the ``ordered()`` method when creating a
``ClientBulkWriteOptions`` object.
```

.. code-block:: java

   MongoNamespace namespace = new MongoNamespace("db", "people");

   ClientBulkWriteOptions options = new ClientBulkWriteOptions().ordered(false);

   ClientBulkWriteResult result = client.bulkWrite(List<>(
      ClientNamespacedWriteModel.insertOne(namespace, new Document("name", "Donkey Kong")),

>       ClientNamespacedWriteModel.insertOne(namespace, new Document("name", "Mario"))

Reviewer 3: S: I think you should select an example that produces an error such as duplicate key or something

      ), 

>       options);

Reviewer 3: S: fix indentation

      { "_id": 6, "name": "Zaynab Omar", "age": 37 }

For more information about the methods and classes mentioned in this section,
see the following API documentation:

- `BulkWriteOptions <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/BulkWriteOptions.html>`__
- `ordered() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/BulkWriteOptions.html#ordered(boolean)>`__

.. _java-sync-client-bulk-write:

Client Bulk Write
-----------------

When connecting to a deployment running {+mdb-server+} 8.0 or later,
you can use the ``MongoClient.bulkWrite()`` method to write

> to multiple databases and collections in the same cluster. The ``MongoClient.bulkWrite()``
> method does not split different kinds of operations into different batches.

Reviewer 2: S: if someone jumps straight to this section, they might not see that this line is intended to compare client bulk writes to collection bulk writes.

```suggestion
to multiple databases and collections in the same cluster. The ``MongoClient.bulkWrite()``
method performs all write operation in a single batch.
```

In the examples in preceding sections of this page, the ``MongoCollection.bulkWrite()`` method 
takes a list of ``WriteModel`` instances in which the specified subclass of 
``WriteModel`` represents the corresponding write operation. For example, an
instance of ``InsertOneModel`` represents an operation to insert one document. 

> Similarly, the ``MongoClient.bulkWrite()`` method takes a 

Reviewer 2: S: unfortunately, i think it's worth repeating the info from earlier in the page just to users don't have to read earlier sections if they don't need to. if you have a lot of repeated content, you can always put it in an `includes` directive 😎

Author: honestly, I'm not sure how much this information is relevant so I will remove it 👍

list of ``ClientNamespacedWriteModel`` instances to represent different write operations.
For example, an instance of ``ClientNamespacedInsertOneModel`` represents an
operation to insert one document.

> You can construct instances of ``ClientNamespacedWriteModel`` using the following

Reviewer 2: ```suggestion
You can construct instances of the ``ClientNamespacedWriteModel`` class by using the following
```

methods:

> - ``insertOne()``

Reviewer 2: S: would use this pattern for each method and delete the following paragraph

```suggestion
- ``ClientNamespacedWriteModel.insertOne()``: Constructs a ``ClientNamespacedInsertOneModel`` instance.
```

- ``updateOne()``

- ``updateMany()``

- ``replaceOne()``

- ``deleteOne()``

- ``deleteMany()``

These methods are used to construct corresponding write models. 
For example, ``ClientNamespacedWriteModel.updateOne()`` is used to
construct a ``ClientNamespacedUpdateOneModel`` instance, which represents an
update operation.

These methods take a ``MongoNamespace`` object that defines which
database and collection to write to. Some, such as ``insertOne()``, can take a 
``Document`` instance that defines information about the write operation. 
Some other methods, such as ``updateOne()`` and ``replaceOne()``, take a filter 

> object that defines the subset of documents to be updated or replaced.

Reviewer 2: for comprehensiveness and clarity, we probably want to specify which methods take what. you could give each method its own subsection, or add a parameter list/table for each one

The following sections provide examples of how to use the client ``bulkWrite()`` 

> method.

Author: Re-review here

When connecting to a deployment running {+mdb-server+} 8.0 or later,
you can use the ``MongoClient.bulkWrite()`` method to write
to multiple databases and collections in the same cluster. The ``MongoClient.bulkWrite()``
method performs all write operations in a single call.

The ``MongoClient.bulkWrite()`` method takes a 
list of ``ClientNamespacedWriteModel`` instances to represent different write operations.
You can construct instances of the ``ClientNamespacedWriteModel`` interface by using 
instance methods. For example, an instance of ``ClientNamespacedInsertOneModel`` represents an
operation to insert one document, and you can create this model by using
the ``ClientNamespacedWriteModel.insertOne()`` method.

The models and their corresponding instance methods are described
in the table below.

.. list-table::
   :header-rows: 1
   
   * - Model
     - Instance Method
     - Description
     - Parameters

   *  - ``ClientNamespacedInsertOneModel``
      - ``insertOne()``
      - Creates a model to insert a document into the ``namespace``.
      - ``namespace``: defines which database and collection to write to
       

>         ``document``: defines the document to insert

Reviewer 2: For all of these parameters, removes 'defines which' or 'defines a' and then capitalize the first letter (e.g., 'Database and collection to write to')

https://www.mongodb.com/docs/meta/style-guide/style/lists/#list-items

   * - ``ClientNamespacedUpdateOneModel``
     - ``updateOne()``

>      - Creates a model to update at most one document in the ``namespace``

Reviewer 2: ```suggestion
     - Creates a model to update the first document in the ``namespace``
```

       that matches ``filter``.
     - ``namespace``: defines which database and collection to write to

       ``filter``: defines filter that selects which document to update

       ``update``: defines update to apply to matching document

       ``updatePipeline``: defines update pipeline to apply to matching document

       ``options``: (optional) defines options to apply when updating document

>        One of ``update`` or ``updatePipeline`` must be specified.

Reviewer 2: ```suggestion
       You must pass a value for either the ``update`` or ``updatePipeline`` parameter.
```

   * - ``ClientNamespacedUpdateManyModel``
     - ``updateMany()``
     - Creates a model to update all documents in the ``namespace`` that match 
       ``filter``.
     - ``namespace``: defines which database and collection to write to

       ``filter``: defines filter that selects which documents to update

       ``update``: defines update to apply to matching documents

       ``updatePipeline``: defines update pipeline to apply to matching documents

       ``options``: (optional) defines options to apply when updating documents

>        One of ``update`` or ``updatePipeline`` must be specified.

Reviewer 2: ```suggestion
       You must pass a value for either the ``update`` or ``updatePipeline`` parameter.
```

   * - ``ClientNamespacedReplaceOneModel``
     - ``replaceOne()``

>      - Creates a model to replace at most one document in the ``namespace`` that 

Reviewer 2: ```suggestion
     - Creates a model to replace the first document in the ``namespace`` that 
```

       matches ``filter``.
     - ``namespace``: defines which database and collection to write to

       ``filter``: defines filter that selects which document to replace

       ``replacement``: defines replacement document

       ``options``: (optional) defines options to apply when replacing documents

   * - ``ClientNamespacedDeleteOneModel``
     - ``deleteOne()``

>      - Creates a model to delete at most one document in the ``namespace`` that 

Reviewer 2: ```suggestion
     - Creates a model to delete the first document in the ``namespace`` that 
```

- Ordered, which performs the bulk operations in order until an error occurs, if any
- Unordered, which performs all the bulk operations in any order and reports errors
  at the end, if any

``MongoClient.bulkWrite()``
~~~~~~~~~~~~~~~~~~~~~~~~~~~

In {+mdb-server+} version 8.0 and {+driver-short+} version 5.3 and later, you 
can use the ``bulkWrite()`` method to perform bulk operations on multiple 
databases and collections at once.

This method uses the ``ClientNamespacedWriteModel`` and its methods
``insertOne()``, ``updateOne()``, ``updateMany()``, ``replaceOne()``, 
``deleteOne()``, and ``deleteMany()``.

The method can also take a ``ClientBulkWriteOptions`` object to specify different
options for how the command is executed.

> .. TODO: Add API Documentation

Reviewer 3: I'd also add this to the whats new for visibility

- Ordered, which performs the bulk operations in order until an error occurs, if any
- Unordered, which performs all the bulk operations in any order and reports errors
  at the end, if any

``MongoClient.bulkWrite()``
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> In {+mdb-server+} version 8.0 and {+driver-short+} version 5.3 and later, you 

Reviewer 3: ```suggestion
When connecting to a deployment running {+mdb-server+} version 8.0 or later, you 
```

> can use the ``bulkWrite()`` method to perform bulk operations on multiple 

Reviewer 3: ```suggestion
can use the ``MongoClient.bulkWrite()`` method to perform bulk operations on multiple 
```

databases and collections at once.

This method uses the ``ClientNamespacedWriteModel`` and its methods
``insertOne()``, ``updateOne()``, ``updateMany()``, ``replaceOne()``, 
``deleteOne()``, and ``deleteMany()``.

The method can also take a ``ClientBulkWriteOptions`` object to specify different
options for how the command is executed.

> To learn more about the client ``bulkWrite`` command, take a look at the 
> :manual:`bulkWrite <reference/command/bulkWrite/>` guide in the {+mdb-server+}
> Manual.

Reviewer 3: ```suggestion
To learn more about the client ``bulkWrite`` command, see the 
:manual:`bulkWrite() <reference/command/bulkWrite/>` method reference in the {+mdb-server+}
Manual.
```

- Ordered, which performs the bulk operations in order until an error occurs, if any
- Unordered, which performs all the bulk operations in any order and reports errors
  at the end, if any

``MongoClient.bulkWrite()``
~~~~~~~~~~~~~~~~~~~~~~~~~~~

When connecting to a deployment running {+mdb-server+} version 8.0 or later, you 
can use the ``MongoClient.bulkWrite()`` method to perform bulk operations on multiple 
databases and collections at once.

This method uses the ``ClientNamespacedWriteModel`` and its methods
``insertOne()``, ``updateOne()``, ``updateMany()``, ``replaceOne()``, 

> ``deleteOne()``, and ``deleteMany()``.

Reviewer 3: S: you can list this list of supported methods in the introduction to client bulk write. maybe around line 325 of the PR instead of only listing 2 methods
